### ISSUE TITLE
Using a struct as an item type instead of a class leads to slient failure in query

### ISSUE THREAD

**Issue by rstat1** [mbdavid/LiteDB#170]
Say I have the following item class:

       public struct ApplicationInfo
       {
              public string InstanceID { get; set; }
              public string ApplicationName { get; set; }
              public ApplicationInfo(string instanceID, string appName)
              {
                    InstanceID = instanceID;
                    ApplicationName = appName;
              }
        }

If I add a bunch of these to a LiteDB collection and then later try to retrieve them through `Collection.Find(Query.All())` all I get is a List of items of the proper type that have all their properties set to null. But if I use the LiteDBShell app to retrieve all the items in the collection, it works as expected.

**Comment by mbdavid** [comment]
Hi @rstat1, structs does support parameters less contructors. So LiteDB can't auto deserialize from BSON document. To serialize is not a problem (so in shell works fine). I suggest to use classes or write your own deserialize function using Register on BsonMapper class.

**Comment by rstat1** [comment]
I knew that :)  I promise.

Thanks for replying though.

**Comment by RenniePet** [comment]
Hi,

I don't understand this. It's true that at the C# source level a struct does not allow a parameterless constructor, but at the CIL level both classes and structs are implemented as CIL classes. I wrote a .Net serializer some years ago, and once I got it working it worked on both C# classes and C# structs without any problem - I don't recall encountering any difference in how they worked using reflection. 

My memory fails me, but fortunately I wrote some notes there three years ago:

---------------------------------------------------------------------------------------

How MCopier instantiates .Net objects for deserialization
---------------------------------------------------------

MCopier uses two different .Net Framework methods for instantiating the objects that it is deserializing: Activator.CreateInstance() and FormatterServices.GetUninitializedObject(). The difference, as implied by the names, is that GetUninitializedObject() does not invoke the object's constructor and does not initialize any of the fields that have initialization specified.

The GetUninitializedObject() method obviously has a performance advantage for most situations involving deserialization, because most usage of a constructor and field initialization is wasted if all of the fields are about to be overwritten with the deserialization data. I.e., both time and storage are saved by using GetUninitializedObject(), and so it is considered the preferred technique. 

However, there can be some unexpected side-effects. The most obvious is that if there is a parameterless constructor that calls methods that have nothing to do with initialization of the object itself then this processing does not get done when GetUninitializedObject() is used. More subtle are the situations that exist if an object has some fields that specify [CopierFormats("")] or [CopierFormats("-x")] (see under "How to use MCopier" and "Attributes" above). When these fields are not involved in the deserialization being done, and if the field definition includes an initialization, then the initialization is not done when GetUninitializedObject() is used by MCopier. The same is true for fields that are not eligible for serialization because they are not declared public, or do not have both get and set accessors, and for XML deserialization this is true for the fields for which no XML element was found (if this is allowed).

The current implementation of MCopier uses FormatterServices.GetUninitializedObject() for all binary deserialization, and for XML deserialization when FCopierOptions.fAllFieldsRequired is turned on. Conversely, it uses Activator.CreateInstance() for XML deserialization when FCopierOptions.fAllFieldsRequired is turned off.

One exception to the above statement is that when deserializing an object that contains a reference to the List<> collection in a System.Collections.ObjectModel.Collection<> base class then the Activator.CreateInstance() method is always used to instantiate the enclosing class. An example of this situation is the KeyedItemList class in the MCollections assembly, which is derived from KeyedCollection

Incidentally, the FCopierOptions.fAllFieldsRequired switch is always turned on for objects processed by methods in MSettings.

---------------------------------------------------------------------------------------------

Yeah, a lot of that is gobbledigook, but you just might get some ideas.
